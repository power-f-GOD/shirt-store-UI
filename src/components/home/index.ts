export * from './BottomBar';
export * from './Card';
