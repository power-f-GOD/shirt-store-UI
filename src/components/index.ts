export * from './Stack';
export * from './Text';
export * from './View';
export * from './Img';
export * from './Skeleton';
