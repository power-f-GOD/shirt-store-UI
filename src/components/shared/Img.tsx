import { memo, useCallback, ReactEventHandler, forwardRef } from 'react';
import Image from 'next/image';

const _Img: typeof Image = forwardRef(({ className, ...props }, ref) => {
  return (
    <Image
      width={360}
      height={400}
      className={`max-w-[75%] h-auto object-cover object-center absolute top-1/2 right-1/2 anim__fadeIn ${
        className || ''
      }`}
      ref={ref}
      {...props}
      alt={props.alt}
      priority
      onError={useCallback<ReactEventHandler>((e) => {
        const img = e.target as HTMLImageElement;
        let sourceExt = '';
        let targetExt = '';

        if (img.src.includes('webp')) {
          sourceExt = 'webp';
          targetExt = 'jpeg';
        } else if (img.src.includes('jpeg')) {
          sourceExt = 'jpeg';
          targetExt = 'png';
        }

        if (!targetExt) return;

        img.srcset = img.srcset.replace(new RegExp(sourceExt, 'gi'), targetExt);
        img.src = img.src.replace(sourceExt, targetExt);
      }, [])}
    />
  );
});

_Img.displayName = 'Img';

export const Img = memo(_Img);
