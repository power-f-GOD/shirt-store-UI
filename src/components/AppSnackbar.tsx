import React, { useCallback, memo } from 'react';
import { shallowEqual } from 'react-redux';
import {
  Snackbar as MuiSnackbar,
  Button,
  Alert,
  Slide,
  SlideProps,
  AlertTitle
} from '@mui/material';

import { dispatch, snackbar, useTypedSelector } from 'src/redux';

const _AppSnackbar = () => {
  const {
    open,
    severity,
    variant,
    message,
    label,
    duration,
    position,
    title,
    autoHide
  } = useTypedSelector((state) => state.snackbar, shallowEqual);

  const handleOnClose = useCallback(() => {
    dispatch(snackbar({ open: false }));
  }, []);

  const SlideTransition = useCallback(
    (props: SlideProps) => {
      return (
        <Slide
          {...props}
          direction={
            position === 'top' || (!position && severity !== 'info')
              ? 'down'
              : 'up'
          }
        />
      );
    },
    [position, severity]
  );

  return (
    <MuiSnackbar
      open={!!open}
      autoHideDuration={duration || 6000}
      onClose={autoHide ? handleOnClose : undefined}
      action={
        <Button color={severity} size="small" onClick={handleOnClose}>
          {label || 'OK'}
        </Button>
      }
      TransitionComponent={
        // severity === 'info' ? undefined :
        SlideTransition
      }
      anchorOrigin={
        position === 'top' || (!position && severity !== 'info')
          ? { vertical: 'top', horizontal: 'right' }
          : globalThis.innerWidth < 768
          ? { vertical: 'bottom', horizontal: 'right' }
          : undefined
      }>
      <Alert
        onClose={handleOnClose}
        severity={severity}
        variant={variant}
        className="shadow-lg">
        {title && (
          <AlertTitle className="text-capitalize mb-1">
            {typeof title === 'boolean' ? `${severity}!` : title}
          </AlertTitle>
        )}
        {message && <>{message[0].toUpperCase() + message.slice(1)}</>}
      </Alert>
    </MuiSnackbar>
  );
};

export const AppSnackbar = memo(_AppSnackbar);
