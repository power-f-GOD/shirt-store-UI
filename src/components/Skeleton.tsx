import { FC, memo, useMemo } from 'react';

import { SkeletonProps } from 'src/types';
import { Stack } from '.';

const _Skeleton: FC<SkeletonProps> = ({
  size,
  width,
  height,
  type,
  borderRadius,
  count,
  variant,
  erred,
  className: _className,
  style: _style
}) => {
  const hasRoundedClass = /(^|\s)rounded/.test(_className || '');
  const hasWidthClass = /(^|\s)w-/.test(_className || '');
  const hasHeightClass = /(^|\s)h-/.test(_className || '');
  const className = `${_className || ''}${
    hasHeightClass || height ? '' : ' h-1m'
  } ${type === 'circle' ? ' rounded-full' : ''}`;
  const style = useMemo(
    () => ({
      ..._style,
      ...((height || size) && !hasHeightClass
        ? { height: height || size }
        : {}),
      ...((width || size) && !hasWidthClass ? { width: width || size } : {}),
      ...(borderRadius && !hasRoundedClass ? { borderRadius } : {})
    }),
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [
      // eslint-disable-next-line react-hooks/exhaustive-deps
      JSON.stringify(_style),
      borderRadius,
      hasHeightClass,
      hasRoundedClass,
      hasWidthClass,
      height,
      size,
      width
    ]
  );

  return (
    <>
      {Array(count || 1)
        .fill('')
        .map((_, i) => (
          <Stack
            key={i}
            as="span"
            className={`Skeleton relative left-0 top-0 inline-flex justify-center items-center self-center max-width-full overflow-hidden rounded-sm ${
              className || ''
            } ${variant || ''} ${erred ? 'erred' : ''}`
              .replace(/\s+/g, ' ')
              .trim()}
            style={style}
          />
        ))}
    </>
  );
};

export const Skeleton = memo(_Skeleton);
