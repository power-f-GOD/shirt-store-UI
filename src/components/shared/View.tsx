import {
  memo,
  createElement,
  forwardRef,
  ForwardRefExoticComponent
} from 'react';
import { ViewProps } from 'src/types';

const _View: ForwardRefExoticComponent<ViewProps> = forwardRef(
  ({ as, children, component, ...props }, ref): JSX.Element | null => {
    const Component = component;

    props.ref = ref;

    if (Component) return <Component {...props}>{children}</Component>;

    return createElement(as || 'div', props, children);
  }
);

_View.displayName = 'View';

export const View = memo(_View);
