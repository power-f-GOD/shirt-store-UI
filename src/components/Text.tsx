import { memo, useMemo, forwardRef } from 'react';

import { TextProps } from 'src/types';
import { View } from './View';

const _Text = forwardRef<HTMLElement, TextProps>(
  (
    {
      as,
      children,
      style,
      className,
      height,
      width,
      background,
      color,
      display,
      fontSize,
      overflow,
      overflowY,
      overflowX,
      alignSelf,
      fontStyle,
      fontWeight,
      overflowEllipsis,
      textAlign,
      textTransform,
      component,
      ...restProps
    },
    ref
  ): JSX.Element => {
    const Component = component || View;
    const styleCheck = JSON.stringify(style || {});

    return (
      <Component
        {...restProps}
        as={as || 'span'}
        ref={ref}
        className={`Text ${className || ''}${
          overflowEllipsis ? ' text-ellipsis' : ''
        }`.trim()}
        style={useMemo(
          () => ({
            alignSelf,
            width,
            height,
            color,
            background,
            display,
            fontSize,
            fontStyle,
            fontWeight,
            overflow,
            overflowY,
            overflowX,
            textAlign,
            textTransform,
            ...(style || {})
          }),
          // eslint-disable-next-line react-hooks/exhaustive-deps
          [
            height,
            width,
            background,
            color,
            display,
            fontSize,
            fontStyle,
            fontWeight,
            overflow,
            overflowX,
            overflowY,
            textAlign,
            textTransform,
            styleCheck
          ]
        )}>
        {children}
      </Component>
    );
  }
);

_Text.displayName = '_Text';

export const Text = memo(_Text);
