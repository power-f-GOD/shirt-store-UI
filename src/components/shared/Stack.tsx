import { memo, forwardRef } from 'react';

import { StackProps } from 'src/types';
import { View } from './View';

const _Stack = forwardRef<HTMLElement, StackProps>(
  (
    {
      children,
      component,
      style,
      className,
      height,
      width,
      align,
      background,
      borderRadius,
      gap,
      color,
      padding,
      direction,
      distribute,
      display,
      zIndex,
      column,
      templateColumns,
      row,
      templateRows,
      columnGap,
      rowGap,
      fontSize,
      opacity,
      overflow,
      overflowY,
      overflowX,
      boxShadow,
      maxHeight,
      minHeight,
      position,
      top,
      right,
      bottom,
      left,
      border,
      borderBottom,
      borderLeft,
      animationDelay,
      animationTimingFunction,
      ...restProps
    },
    ref
  ): JSX.Element => {
    const Component = component || View;

    return (
      <Component
        {...restProps}
        ref={ref}
        className={`Stack flex${
          !className?.includes('flex-row') ? ' flex-col' : ''
        } ${className || ''}`}
        style={{
          alignItems: align,
          width,
          height,
          alignContent: align,
          background,
          borderRadius,
          gap,
          color,
          padding,
          zIndex,
          display,
          flexDirection: direction,
          justifyContent: distribute,
          gridRow: row,
          gridTemplateRows: templateRows,
          gridColumn: column,
          gridTemplateColumns: templateColumns,
          columnGap: columnGap || gap,
          rowGap: rowGap || gap,
          boxShadow,
          fontSize,
          opacity,
          maxHeight,
          minHeight,
          position,
          top,
          right,
          bottom,
          left,
          overflow,
          overflowY,
          overflowX,
          borderBottom: borderBottom || border,
          borderTop: border,
          borderLeft: borderLeft || border,
          borderRight: border,
          animationDelay,
          animationTimingFunction,
          ...(style || {})
        }}>
        {children}
      </Component>
    );
  }
);

_Stack.displayName = '_Stack';

export const Stack = memo(_Stack);
