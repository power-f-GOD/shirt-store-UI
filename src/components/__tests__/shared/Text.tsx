import { render } from '@testing-library/react';

import { Text } from 'src/components/shared';

it('renders the Text component unchanged', () => {
  const { container } = render(<Text />);

  expect(container).toMatchSnapshot();
});
