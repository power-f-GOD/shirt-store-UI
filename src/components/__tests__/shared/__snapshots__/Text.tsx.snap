// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the Text component unchanged 1`] = `
<div>
  <span
    class="Text"
  />
</div>
`;
